pkg/tcpip/iptables/iptables.go

```diff
@@ -184,8 +185,16 @@ func (it *IPTables) checkTable(hook Hook, pkt tcpip.PacketBuffer, tablename stri
 	panic(fmt.Sprintf("Traversed past the entire list of iptables rules in table %q.", tablename))
 }
 
+// Precondition: pk.NetworkHeader is set.
 func (it *IPTables) checkRule(hook Hook, pkt tcpip.PacketBuffer, table Table, ruleIdx int) Verdict {
 	rule := table.Rules[ruleIdx]
+
+	// First check whether the packet matches the IP header filter.
+	// TODO(gvisor.dev/issue/170): Support other fields of the filter.
+	if rule.Filter.Protocol != 0 && rule.Filter.Protocol != header.IPv4(pkt.NetworkHeader).TransportProtocol() {
```

Comment:
Is the `!= 0` check necessary?  Can `TransportProtocol()` return 0?

Reply:
TransportProtocol() can return 0 -- there's nothing stopping it. rule.Filter.Protocol == 0 indicates that all protocols are matched. The iptables man page discusses this, and that you can't use `-p 0` to test for a 0 value transport protocol number.